include/common/nas_structs.h

```diff
@@ -0,0 +1,347 @@
+#define ESM_MSG_ACTV_DEF_BEAR__CTX_REQ 0xc1
+
+#define AUTH_RESPONSE 53
+#define AUTH_SYNC_FAILURE 21
+
+#define MAX_NAS_MSG_SIZE 500
+
+/*TODO: Needed fro paging. Ignoring for now*/
+struct proto_conf {
+    int placeholder;
+};
+struct esm_sec_info {
+    struct proto_conf proto_config;
+};
+
+/* Should we put this in NAS header ?*/
+/* sec. 10.5.6.3 of 3GPP TS 24.008 has following statement:
+ * The protocol configuration options is a type 4 information element with a minimum length of 3 octets and a maximum length of 253 octets.
+ */
+#define MAX_PCO_OPTION_SIZE 255
+struct pco 
+{
+    unsigned char pco_length;
+    unsigned char pco_options[MAX_PCO_OPTION_SIZE];
+};
+
+typedef struct pdn_address {
+    uint8_t spare :5;
+    uint8_t type :3;
+    uint32_t ipv4; /* TODO: Revisit 24.301 - 9.9.4.9.1 */
+} pdn_address;
+
+typedef struct linked_transcation_id {
+    uint8_t flag :1;
+    uint8_t val :7;
+} linked_transcation_id;
+
+enum protocol_discriminator{
+    EPSSessionManagementMessage = 2,
+    EPSMobilityManagementMessages = 7,
+};
+
+typedef struct esm_qos
+{
+    uint8_t reliability_class :3;
+    uint8_t delay_class :3;
+    uint8_t spare1 :2;
+    uint8_t precedence_class :3;
+    uint8_t spare2 :1;
+    uint8_t peak_throughput :4;
+    uint8_t mean_throughput :5;
+    uint8_t spare3 :3;
+    uint8_t delivery_err_sdu :3;
+    uint8_t delivery_order :2;
+    uint8_t traffic_class :3;
+    uint8_t max_sdu_size;
+    uint8_t mbr_ul;
+    uint8_t mbr_dl;
+    uint8_t sdu_err_ratio :4;
+    uint8_t residual_ber :4;
+    uint8_t trffic_prio :2;
+    uint8_t transfer_delay :6;
+    uint8_t gbr_ul;
+    uint8_t gbr_dl;
+    uint8_t src_stat_desc :4;
+    uint8_t sig_ind :1;
+    uint8_t spare4 :3;
+    uint8_t mbr_dl_ext;
+    uint8_t gbr_dl_ext;
+    uint8_t mbr_ul_ext;
+    uint8_t gbr_ul_ext;
+
+} esm_qos;
+
+/* TODO : Revisit 24.301 - 9.9.4.2.1 */
+typedef struct ESM_APN_AMBR {
+    uint8_t dl;
+    uint8_t reserved;
+    uint8_t dlext;
+    uint8_t ulext;
+    uint8_t dlext2;
+    /* uint8_t dl_total; */
+    uint8_t ulext2;
+    /* uint8_t ul_total; */
+} ESM_APN_AMBR;
+
+#define AUTHREQ_NAS_SECURITY_PARAM 0x01
+#define SERVICE_REQ_SECURITY_HEADER 12
+
+#define BINARY_IMSI_LEN 8 /*same as packet capture. TODO: Write macros*/
+
+/* NAS Security Encryption Algorithm */
+typedef enum security_encryption_algo {
+    Algo_EEA0 = 0,
+
+}security_encryption_algo;
+
+/* NAS Security Integrity Algorithm */
+typedef enum security_integrity_algo {
+    Algo_EIA0 = 0,
+    Algo_128EIA1 = 1,
+}security_integrity_algo;
+
+
+/* s1apCauseMisc */
+typedef long     s1apCauseMisc_t;
+/* s1apCauseProtocol */
+typedef long     s1apCauseProtocol_t;
+/* s1apCauseNas */
+typedef long     s1apCauseNas_t;
+/* s1apCauseTransport */
+typedef long     s1apCauseTransport_t;
+/* s1apCauseRadioNetwork */
+typedef long     s1apCauseRadioNetwork_t;
+
+
+
+typedef enum emmCause {
+    emmCause_ue_id_not_derived_by_network   = 9,
+    emmCause_network_failure = 11
```

Comment:
network failure cause should be 17.
Need to move nasDetachType enum as well from s1ap_structs.h
(changes in the commit for erab modification)

Reply:
Good catch. Fixed this.